neural_speed/models/llama/llama_yarn_utils.cpp

```diff
@@ -0,0 +1,200 @@
+                         void* progress_callback_user_data) {
+  std::unique_ptr<Llama> ms(new Llama());
+  ms->init(fname.c_str(), ctx, n_gpu_layers, use_mmap, use_mlock, vocab_only);
+  ms->load(ctx, progress_callback, progress_callback_user_data);
+
+  model_context& lctx = *ctx;
+  lctx.support_bestla_kv = true;
+}
+
+void Llama::init(const char* path_model, model_context* ctx, int n_gpu_layer_, bool use_mmap_, bool use_mlock_,
+                 bool vocab_only_) {
+  model_context& lctx = *ctx;
+  n_gpu_layer = n_gpu_layer_;
+  use_mmap = use_mmap_;
+  use_mlock = use_mlock_;
+  vocab_only = vocab_only_;
+  auto& model = lctx.model;
+  ml.reset(new model_model_loader(path_model, use_mmap, vocab_only));
+  lctx.vocab = std::move(ml->file_loaders.at(0)->vocab);
+  model.hparams = ml->file_loaders.at(0)->hparams;
+  model_file_version file_version = ml->file_loaders.at(0)->file_version;
+  auto& hparams = model.hparams;
+  n_ff = hparams.n_mult;
+  fprintf(stderr, "%s: n_vocab    = %u\n", __func__, hparams.n_vocab);
+  fprintf(stderr, "%s: n_ctx      = %u\n", __func__, hparams.max_seq_len);
+  fprintf(stderr, "%s: n_embd     = %u\n", __func__, hparams.n_embd);
+  fprintf(stderr, "%s: n_mult     = %u\n", __func__, hparams.n_mult);
+  fprintf(stderr, "%s: n_head     = %u\n", __func__, hparams.n_head);
+  fprintf(stderr, "%s: n_head_kv  = %u\n", __func__, hparams.n_head_kv);
+  fprintf(stderr, "%s: n_layer    = %u\n", __func__, hparams.n_layer);
+  fprintf(stderr, "%s: n_rot      = %u\n", __func__, hparams.n_rot);
+  fprintf(stderr, "%s: n_ff       = %u\n", __func__, n_ff);
+  fprintf(stderr, "%s: n_parts    = %zu\n", __func__, ml->file_loaders.size());
+  n_embd = hparams.n_embd;
+  n_vocab = hparams.n_vocab;
+  n_layer = hparams.n_layer;
+  n_head_kv = hparams.n_head_kv;
+  n_head = hparams.n_head;
+  scratch = llama_mem_req(n_layer);
+  model.scratchs = scratch;
+}
+
+#define MODEL_BACKEND_OFFLOAD NE_BACKEND_CPU
+void Llama::load(model_context* ctx, model_progress_callback progress_callback, void* progress_callback_user_data) {
+  model_context& lctx = *ctx;
+  auto& model = lctx.model;
+  auto& ne_ctx = model.ctx;
+  size_t ctx_size;
+  size_t mmapped_size;
+  ml->calc_sizes(&ctx_size, &mmapped_size);
+  fprintf(stderr, "%s: ne ctx size = %7.2f MB\n", __func__, ctx_size / 1024.0 / 1024.0);
+
+  // create the ne context
+  lctx.model.buf.resize(ctx_size);
+  if (use_mlock) {
+    lctx.model.mlock_buf.init(lctx.model.buf.addr);
+    lctx.model.mlock_buf.grow_to(lctx.model.buf.size);
+  }
+
+  struct ne_init_params params = {
+      /*.mem_size   =*/lctx.model.buf.size,
+      /*.mem_buffer =*/lctx.model.buf.addr,
+      /*.no_alloc   =*/ml->use_mmap,
+  };
+
+  model.ctx = ne_init(params);
+  if (!model.ctx) {
+    throw format("ne_init() failed");
+  }
+
+  ml->ne_ctx = ne_ctx;
+
+  const int i_gpu_start = n_layer - n_gpu_layer;
+  model.layers.resize(n_layer);
+  size_t vram_total = 0;
+  if (1) {
```

Comment:
why use if(1)?

Reply:
Should be fixed. 
The PR is split. The code is deleted now. Should be fixed and submitted in next PR.